File: lib/data/repository/contract/vitals_repository.dart
import '../../../core/period/period_load_query.dart';
import '../../../core/result/result.dart';
import '../../../core/time/local_date.dart';
import '../../../domain/model/refresh_mode.dart';
import '../../../domain/model/vitals_models.dart';
import '../../../domain/query/vitals_period_data.dart';

/// Which vitals metric family a period load should populate. Port of the Kotlin
/// `VitalsPeriodMetric`.
enum VitalsPeriodMetric {
  all,
  bloodPressure,
  spo2,
  vo2Max,
  respiratoryRate,
  bodyTemperature,
  bloodGlucose,
  skinTemperature,
}

/// Port of the Kotlin `VitalsRepository` contract.
///
/// Fallible operations return [Result]; the synchronous probes
/// ([phase3Permissions], [vitalsWritePermissions]) read cached state and
/// cannot fail, so they stay bare.
abstract interface class VitalsRepository {
  Set<String> get phase3Permissions;

  Set<String> vitalsWritePermissions(VitalsMeasurementType type);

  Future<Result<Set<String>>> missingPermissions();

  Future<Result<VitalsPeriodData>> loadVitalsPeriod(
    PeriodLoadQuery query,
    VitalsPeriodMetric metric, {
    RefreshMode refreshMode = RefreshMode.normal,
  });

  Future<Result<List<BloodPressureEntry>>> loadBloodPressure(
    LocalDate start,
    LocalDate end,
  );

  Future<Result<List<SpO2Entry>>> loadSpO2(LocalDate start, LocalDate end);

  Future<Result<List<RespiratoryRateEntry>>> loadRespiratoryRate(
    LocalDate start,
    LocalDate end,
  );

  Future<Result<List<BodyTempEntry>>> loadBodyTemperature(
    LocalDate start,
    LocalDate end,
  );

  Future<Result<List<Vo2MaxEntry>>> loadVo2Max(LocalDate start, LocalDate end);

  Future<Result<List<BloodGlucoseEntry>>> loadBloodGlucose(
    LocalDate start,
    LocalDate end,
  );

  Future<Result<List<SkinTemperatureEntry>>> loadSkinTemperature(
    LocalDate start,
    LocalDate end,
  );

  Future<Result<bool>> hasVitalsWritePermission(VitalsMeasurementType type);

  Future<Result<String>> writeVitalsMeasurementEntry(
    VitalsMeasurementWriteRequest request,
  );

  Future<Result<VitalsMeasurementEntry?>> loadVitalsMeasurementEntry(
    VitalsMeasurementType type,
    String id,
  );

  Future<Result<void>> updateVitalsMeasurementEntry(
    String id,
    VitalsMeasurementWriteRequest request,
  );

  Future<Result<void>> deleteVitalsMeasurementEntry(
    VitalsMeasurementType type,
    String id,
  );
}

File: app/src/main/kotlin/tech/mmarca/openvitals/data/repository/contract/VitalsRepository.kt
package tech.mmarca.openvitals.data.repository.contract

import java.time.LocalDate
import tech.mmarca.openvitals.core.period.PeriodLoadQuery
import tech.mmarca.openvitals.data.repository.VitalsPeriodMetric
import tech.mmarca.openvitals.domain.model.BloodGlucoseEntry
import tech.mmarca.openvitals.domain.model.BloodPressureEntry
import tech.mmarca.openvitals.domain.model.DailyBloodPressurePoint
import tech.mmarca.openvitals.domain.model.DailyVitalPoint
import tech.mmarca.openvitals.domain.model.BodyTempEntry
import tech.mmarca.openvitals.domain.model.RefreshMode
import tech.mmarca.openvitals.domain.model.RespiratoryRateEntry
import tech.mmarca.openvitals.domain.model.SkinTemperatureEntry
import tech.mmarca.openvitals.domain.model.SpO2Entry
import tech.mmarca.openvitals.domain.model.VitalsMeasurementEntry
import tech.mmarca.openvitals.domain.model.VitalsMeasurementType
import tech.mmarca.openvitals.domain.model.VitalsMeasurementWriteRequest
import tech.mmarca.openvitals.domain.model.Vo2MaxEntry
import tech.mmarca.openvitals.domain.query.VitalsPeriodData

interface VitalsRepository {
    val phase3Permissions: Set<String>

    fun vitalsWritePermissions(type: VitalsMeasurementType): Set<String>

    suspend fun missingPermissions(): Set<String>

    suspend fun loadVitalsPeriod(
        query: PeriodLoadQuery,
        metric: VitalsPeriodMetric,
        refreshMode: RefreshMode = RefreshMode.NORMAL,
    ): VitalsPeriodData

    /**
     * One aggregated point per local day for a single vitals metric — the
     * overview's per-metric read, exposed for range consumers such as the
     * health report. Served from the daily cache when the sync cursor covers
     * the range, from a live read otherwise. [metric] must name a concrete
     * vitals metric: ALL and BLOOD_PRESSURE (whose daily shape carries two
     * values, see [loadDailyBloodPressure]) are rejected.
     */
    suspend fun loadDailyVitals(
        metric: VitalsPeriodMetric,
        start: LocalDate,
        end: LocalDate,
    ): List<DailyVitalPoint>

    /** Daily systolic/diastolic averages, cache-first like [loadDailyVitals]. */
    suspend fun loadDailyBloodPressure(start: LocalDate, end: LocalDate): List<DailyBloodPressurePoint>

    suspend fun loadBloodPressure(start: LocalDate, end: LocalDate): List<BloodPressureEntry>

    suspend fun loadSpO2(start: LocalDate, end: LocalDate): List<SpO2Entry>

    suspend fun loadRespiratoryRate(start: LocalDate, end: LocalDate): List<RespiratoryRateEntry>

    suspend fun loadBodyTemperature(start: LocalDate, end: LocalDate): List<BodyTempEntry>

    suspend fun loadVo2Max(start: LocalDate, end: LocalDate): List<Vo2MaxEntry>

    suspend fun loadBloodGlucose(start: LocalDate, end: LocalDate): List<BloodGlucoseEntry>

    suspend fun loadSkinTemperature(start: LocalDate, end: LocalDate): List<SkinTemperatureEntry>

    suspend fun hasVitalsWritePermission(type: VitalsMeasurementType): Boolean

    suspend fun writeVitalsMeasurementEntry(request: VitalsMeasurementWriteRequest): String

    suspend fun loadVitalsMeasurementEntry(type: VitalsMeasurementType, id: String): VitalsMeasurementEntry?

    suspend fun updateVitalsMeasurementEntry(id: String, request: VitalsMeasurementWriteRequest)

    suspend fun deleteVitalsMeasurementEntry(type: VitalsMeasurementType, id: String)
}
